src/NodaTime/DateInterval.cs

```diff
@@ -167,5 +193,51 @@ public void Deconstruct(out LocalDate start, out LocalDate end)
             start = Start;
             end = End;
         }
+
+        /// <summary>
+        /// Returns the intersection between the given interval and this interval.
+        /// </summary>
+        /// <param name="interval">
+        /// The specified interval to which return the intersection with.
+        /// </param>
+        /// <returns>
+        /// A <see cref="DateInterval"/> corresponding to the intersection between the given interval and the current
+        /// instance. If there is no intersection, a null reference is returned.
+        /// </returns>
+        /// <exception cref="ArgumentException">
+        /// The start and end dates of <paramref name="interval" /> are not in the same calendar
+        /// as the start and end date of this interval.
+        /// </exception>
+        [CanBeNull]
+        public DateInterval Intersection([NotNull]DateInterval interval)
+        {
+            ValidateInterval(interval);
+
+            if (interval == this || Contains(interval))
```

Comment:
Nit: could we have braces around these conditions? Or possibly a big set of conditional expressions:

```csharp
return interval == this || Contains(interval) ? interval
    : interval.Contains(this) ? this
    : interval.Contains(this.Start) ? new DateInterval(interval.Start, End)
    : interval.Contains(this.End) ? new DateInterval(Start, interval.End)
    : null;
```